### ISSUE TITLE
Load pretrained model

### ISSUE THREAD

**Issue by devguy83** [long123524/BsiNet-torch#10]
Hello 
When I load pretrained model, i got an issue , as you can see in attached file. 
![image](https://github.com/long123524/BsiNet-torch/assets/54128191/204d9a7d-d9aa-4ae2-a628-6f8af8321450)
So i modified test.py as like below   
![image](https://github.com/long123524/BsiNet-torch/assets/54128191/01dc9130-4e87-4558-94a0-c21252965d89)
so i could fix tha issue,but i got different inference result whenever I run inference with that model.
How can i fix it?

**Comment by long123524** [comment]
> Hello When I load pretrained model, i got an issue , as you can see in attached file. ![image](https://private-user-images.githubusercontent.com/54128191/327693934-204d9a7d-d9aa-4ae2-a628-6f8af8321450.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ3MjY2MTgsIm5iZiI6MTcxNDcyNjMxOCwicGF0aCI6Ii81NDEyODE5MS8zMjc2OTM5MzQtMjA0ZDlhN2QtZDlhYS00YWUyLWE2MjgtNmY4YWY4MzIxNDUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAzVDA4NTE1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRhN2NkN2E4OTIwYmQyMTdjMzUyNjE3OWZlYTM2NDMzY2MxOGMxMjMyMWU2NTQ4MzI4ODQyN2VjNGYwMjg5NmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cQHZwEVxWqF5VvWuRDqQghIRH7ZsmlFWn1pVfBfO2Is) So i modified test.py as like below ![image](https://private-user-images.githubusercontent.com/54128191/327694112-01dc9130-4e87-4558-94a0-c21252965d89.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ3MjY2MTgsIm5iZiI6MTcxNDcyNjMxOCwicGF0aCI6Ii81NDEyODE5MS8zMjc2OTQxMTItMDFkYzkxMzAtNGU4Ny00NTU4LTk0YTAtYzIxMjUyOTY1ZDg5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAzVDA4NTE1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJhYWY4ZDc5ZTlhYzUyMzU4NTRiMzVhMmE1YmU3NWI1ZjUyODRlMjM3OWE1YmExNWVkOWMyMzYyNzUyOTRkNzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kJGwBGcugnlkKQGSV6cKYaSekT-25ef72Q5DZ3PtPWU) so i could fix tha issue,but i got different inference result whenever I run inference with that model. How can i fix it?

you maybe use model.load_state_dict(torch.load(model_file), Trure)  or comment model=nn.DataParallel(model). You can try it. Thanks!

**Comment by long123524** [comment]
> Hello When I load pretrained model, i got an issue , as you can see in attached file. ![image](https://private-user-images.githubusercontent.com/54128191/327693934-204d9a7d-d9aa-4ae2-a628-6f8af8321450.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ3Mjc0OTEsIm5iZiI6MTcxNDcyNzE5MSwicGF0aCI6Ii81NDEyODE5MS8zMjc2OTM5MzQtMjA0ZDlhN2QtZDlhYS00YWUyLWE2MjgtNmY4YWY4MzIxNDUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAzVDA5MDYzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwNTAwMzNhZWU1NDZlZTFmZmFiNzcwYWFjOWVkZmUyOGM0MzlhNmYyNGNkZTQzMWNmNTAzODg4YmE3MWU2YzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1DfbTNRdbpjXupDcjpdjcaOs4YlpoXs8RNP4vPfXqUM) So i modified test.py as like below ![image](https://private-user-images.githubusercontent.com/54128191/327694112-01dc9130-4e87-4558-94a0-c21252965d89.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ3Mjc0OTEsIm5iZiI6MTcxNDcyNzE5MSwicGF0aCI6Ii81NDEyODE5MS8zMjc2OTQxMTItMDFkYzkxMzAtNGU4Ny00NTU4LTk0YTAtYzIxMjUyOTY1ZDg5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAzVDA5MDYzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxZmVlMjBjZGIzMzIyOTdmZDVjMzIxZDk2Mzk0MTlmNWViM2Y2NjEzYWMwOTI1NDVhZDZmZjM4YWQ2MmQ3MGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hjHeDPwHH7lPVPjYrLX-uUrK9fBJWPt1fHDO3tAXhXU) so i could fix tha issue,but i got different inference result whenever I run inference with that model. How can i fix it?

Please consider the following questions: 1. The pre-trained weights provided are based on the GPU. 2. The code "model=nn.DataParallel(model)" are also based on the GPU. 3. The impact of the version of different torch. You maybe use model.load_state_dict(torch.load(model_file), False) to run it or comment mdoel=nn.DataParallel(model). Thanks!

**Comment by devguy83** [comment]
I am inferencing on cpu.
whenever inferencing with same image , the result is not same
![image](https://github.com/long123524/BsiNet-torch/assets/54128191/638f3ab6-5e42-4f64-8088-3dc7b5619844)
it looks that the bias values was missed from pretrained weight

**Comment by devguy83** [comment]
As you can see the issue raised when loadig the model, it tells bias values are not in the model
what's the reason? I have tried with 1024 x 1024 size  image.

**Comment by devguy83** [comment]
Second question is how to set the params for normalizing the image. 
ex :  transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])

**Comment by long123524** [comment]
> Second question is how to set the params for normalizing the image. ex : transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])

you can calculate the mean and standard deviation according your dataset. In this practice, this seems to have had very little impact.

**Comment by long123524** [comment]
> As you can see the issue raised when loadig the model, it tells bias values are not in the model what's the reason? I have tried with 1024 x 1024 size image.

For this strange problem, I have not a well explantion, maybe there is an equipment or environmental deviation here? . You can try to use the GPU to do predict.  Besides, you can try to train this model from scratch using our provided dataset, and then predict results.

**Comment by devguy83** [comment]
if it's reason from device or environment problem, you can modify a bit your code to inference it on cpu.
Maybe you should have same issue like me.
I think that bias weight was missed in pre-trained model due to unknown reason.

**Comment by long123524** [comment]
> if it's reason from device or environment problem, you can modify a bit your code to inference it on cpu. Maybe you should have same issue like me. I think that bias weight was missed in pre-trained model due to unknown reason.

Thanks. I did find the same question. I will train the model to get a new weight. I will make sure the weight is available before uploading.

**Comment by long123524** [comment]
> if it's reason from device or environment problem, you can modify a bit your code to inference it on cpu. Maybe you should have same issue like me. I think that bias weight was missed in pre-trained model due to unknown reason.

We have updated the pre-trained weight of BsiNet and successfully tested it. you can download from https://drive.google.com/drive/folders/121T8FjiyEsIbfyLUbrBXYCg75PIzCzRX?usp=sharing, and used it by comment the code "model=nn.DataParallel(model)". Thanks for your concern.

**Comment by devguy83** [comment]
I have checked out with uploaded model,now it works well including loading model and consistens inference result .
Thank you for your help.
